components/calendar/src/persian.rs

```diff
@@ -0,0 +1,921 @@
+        } else {
+            ceil_div(day_of_year - 6, 30) as u8
+        };
+        #[allow(clippy::unwrap_used)] // month and day
+        let day = (date - Self::fixed_from_persian_integers(year, month, 1).unwrap() + 1) as u8;
+        #[allow(clippy::unwrap_used)] // valid month and day
+        Date::try_new_persian_date(year, month, day).unwrap()
+    }
+
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L4829
+    fn arithmetic_persian_year_from_fixed(date: RataDie) -> i64 {
+        #[allow(clippy::unwrap_used)] // valid year,month,day
+        let d0 = date - Self::fixed_from_persian_integers(475, 1, 1).unwrap();
+        let d = div_rem_euclid64(d0, 1029983);
+        let n2820 = d.0;
+        let d1 = d.1;
+        let y2820 = if d1 == 1029982 {
+            2820
+        } else {
+            div_rem_euclid64(128 * d1 + 46878, 46751).0
+        };
+        let year = 474 + n2820 * 2820 + y2820;
+        if year > 0 {
+            year
+        } else {
+            year - 1
+        }
+    }
+
+    pub(crate) fn fixed_from_persian_integers(year: i32, month: u8, day: u8) -> Option<RataDie> {
+        Date::try_new_persian_date(year, month, day)
+            .ok()
+            .map(|d| *d.inner())
+            .map(Self::fixed_from_arithmetic_persian)
+    }
+
+    fn days_in_year_direct(year: i32) -> u32 {
+        if Persian::is_leap_year(year) {
+            366
+        } else {
+            365
+        }
+    }
+
+    fn year_as_persian(year: i32) -> types::FormattableYear {
+        if year > 0 {
+            types::FormattableYear {
+                era: types::Era(tinystr!(16, "ah")),
+                number: year,
+                related_iso: None,
+            }
+        } else {
+            types::FormattableYear {
+                era: types::Era(tinystr!(16, "bh")),
+                number: year,
+                related_iso: None,
+            }
+        }
+    }
+}
+
+impl Date<Persian> {
+    /// Construct new Persian Date.
+    ///
+    /// Has no negative years, only era is the AH/AP.
+    ///
+    /// ```rust
+    /// use icu::calendar::Date;
+    ///
+    /// let date_persian = Date::try_new_persian_date(1392, 4, 25)
+    ///     .expect("Failed to initialize Persian Date instance.");
+    ///
+    /// assert_eq!(date_persian.year().number, 1392);
+    /// assert_eq!(date_persian.month().ordinal, 4);
+    /// assert_eq!(date_persian.day_of_month().0, 25);
+    /// ```
+    pub fn try_new_persian_date(
+        year: i32,
+        month: u8,
+        day: u8,
+    ) -> Result<Date<Persian>, CalendarError> {
+        let inner = ArithmeticDate {
+            year,
+            month,
+            day,
+            marker: PhantomData,
+        };
+
+        let bound = inner.days_in_month();
+        if day > bound {
+            return Err(CalendarError::OutOfRange);
+        }
+        Ok(Date::from_raw(PersianDateInner(inner), Persian))
```

Comment:
Issue: You aren't checking that `month` is in range. See #3543 for a cleaner solution for this.